File: Data Structure And Algorithms/Recursion/merge sort.cpp
#include <bits/stdc++.h>
using namespace std;
vector<int> merge(vector<int> left, vector<int> right)
{
    int i = 0, j = 0, k = 0;
    vector<int> combine(left.size() + right.size());
    while (i < left.size() and j < right.size())
    {
        if (left[i] < right[j])
        {
            combine[k++] = left[i++];
        }
        else
        {
            combine[k++] = right[j++];
        }
    }
    while (i < left.size())
    {
        combine[k++] = left[i++];
    }
    while (j < right.size())
    {
        combine[k++] = right[j++];
    }
    return combine;
}
vector<int> mergeSort(vector<int> v)
{
    if (v.size() == 1)
        return v;
    int mid = v.size() / 2;
    vector<int> left = mergeSort({v.begin(), v.begin() + mid}); // passing copy of vector
    vector<int> right = mergeSort({v.begin() + mid, v.end()});
    return merge(left, right);
}
int main()
{
    vector<int> v = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
    v = mergeSort(v);
    for (auto i : v)
        cout << i << " ";
    return 0;
}

File: Java/MergeSort.java
public class MergeSort {
    public static void main(String[] args) {
        int[] arr={15,98,66,24,76,33,19,82,10,50,75,48};
        mergeSort(arr,0,arr.length-1);
        printArray(arr,arr.length);

    }
    public static void merge(int a[], int beg, int mid, int end)
{
    int i, j, k;
    int n1 = mid - beg + 1;
    int n2 = end - mid;

    /* temporary Arrays */
    int[] LeftArray = new int[n1];
    int[] RightArray = new int[n2];

    /* copy data to temp arrays */
    for (i = 0; i < n1; i++)
        LeftArray[i] = a[beg + i];
    for (j = 0; j < n2; j++)
        RightArray[j] = a[mid + 1 + j];

    i = 0; /* initial index of first sub-array */
    j = 0; /* initial index of second sub-array */
    k = beg;  /* initial index of merged sub-array */

    while (i < n1 && j < n2)
    {
        if(LeftArray[i] <= RightArray[j])
        {
            a[k] = LeftArray[i];
            i++;
        }
        else
        {
            a[k] = RightArray[j];
            j++;
        }
        k++;
    }
    while (i<n1)
    {
        a[k] = LeftArray[i];
        i++;
        k++;
    }

    while (j<n2)
    {
        a[k] = RightArray[j];
        j++;
        k++;
    }
}

   public static void mergeSort(int a[], int beg, int end)
    {
        if (beg < end)
        {
            int mid = (beg + end) / 2;
            mergeSort(a, beg, mid);
            mergeSort(a, mid + 1, end);
            merge(a, beg, mid, end);
        }
    }

  public static void printArray(int[] a, int n)
    {
        int i;
        for (i = 0; i < n; i++)
            System.out.print(a[i] + " ");
    }
}
